src/catalog/src/durable/impls/persist.rs

```diff
@@ -362,6 +418,43 @@ impl UnopenedPersistCatalogState {
             txn.commit().await?;
         }
 
+        // Now that we've fully opened the catalog at the current version, we can increment the
+        // version in the upgrade shard to signal to readers that the allowable versions have
+        // increased.
+        if matches!(mode, Mode::Writable) {
+            let upgrade_shard_id = Self::shard_id(self.organization_id, UPGRADE_SEED);
+            let mut write_handle: WriteHandle<(), (), Timestamp, Diff> = catalog
```

Comment:
I'd personally pull this whole open_writer/compare_and_append bit out into a helper method

Reply:
Done.